File: src/revgate/cli/review_command.py
"""The `revgate review` command.

NOT a hook: no payload on stdin, no hook JSON on stdout, real exit codes. The
plan half reuses `gate_plan` from the hook module.
"""

import os
import time
from pathlib import Path

from revgate.cli.grammar import CliOptions
from revgate.cli.plan_hook import ReviewOutcome, gate_plan
from revgate.git.collect import collect_diff
from revgate.git.scope import ScopeError, describe_scope, filter_files
from revgate.git.staging import get_stage_states
from revgate.review.diff import parse_unified_diff
from revgate.review.report import ReviewOutcomeSummary, review_report
from revgate.server.app import ReviewContext, start_review_server
from revgate.server.browser import open_browser
from revgate.server.wsgi import ServerClosed
from revgate.shared.log import log, warn
from revgate.shared.streams import write_stdout
from revgate.shared.types import DiffFile, HookPayload
from revgate.store.history import HistoryMeta, save_history


def resolve_plan(options: CliOptions, cwd: str) -> str | None:
    """The plan text to review, or None when `--plan` was not asked for.

    `--plan <file>` beats `$REVGATE_PLAN_FILE`. Strict, because the skill reads
    exit 0 as "approved, start implementing" — a missing plan is bad usage.
    """
    if not options.plan:
        return None

    file = options.plan_file or os.environ.get("REVGATE_PLAN_FILE")
    if file:
        try:
            text = (Path(cwd) / file).read_text(encoding="utf-8", errors="replace")
        except OSError as err:
            raise ScopeError(f"could not read plan file {file}: {err}") from err
        # An existing-but-empty file is not a plan: reviewing a blank document
        # the reviewer can only approve is a sign-off on nothing.
        if text.strip():
            return text
        warn(f"plan file {file} is empty")
    raise ScopeError(
        "--plan was given but no plan text was found — pass a file or set $REVGATE_PLAN_FILE"
    )


def _deliver(text: str, options: CliOptions, cwd: str) -> None:
    """`--output` when asked, otherwise stdout.

    stdout is the only thing this command ever writes there, and never hook JSON.
    """
    if not options.output:
        write_stdout(text)
        return
    destination = Path(cwd) / options.output
    try:
        destination.write_text(text, encoding="utf-8", newline="")
    except OSError as err:
        # The human has already reviewed, and exit 1 reads as "no verdict".
        warn(f"could not write {destination}: {err}")
        warn("writing the annotations to stdout instead")
        write_stdout(text)
        return
    log(f"annotations written to {destination}")


def run_review_command(options: CliOptions) -> int:
    """`revgate review` — the on-demand entry point the skill drives.

    Returns the exit code; it never sets it itself.
    """
    cwd = str(Path.cwd())
    payload = HookPayload(session_id="cli", timestamp=int(time.time() * 1000), cwd=cwd)

    plan_text = resolve_plan(options, cwd)
    outcome = (
        gate_plan(payload, plan_text, options)
        if plan_text is not None
        else review_diff(payload, options)
    )

    report = review_report(
        outcome.summary, "plan" if plan_text is not None else "diff", options.exit_code_on_comments
    )
    if report.kind == "interrupted":
        warn("no verdict was captured — reporting an error rather than an approval")
    elif report.kind == "not-a-repo":
        warn("not a git repository — nothing was reviewed")
        warn("run `revgate review` from inside a repository")
    elif report.kind == "scan-failed":
        warn("the untracked-file scan failed — reporting an error rather than an approval")
    elif report.kind == "dropped-paths":
        warn("every changed file was dropped for a line break in its path — not an approval")
    _deliver(report.text, options, cwd)
    return report.exit_code


def review_diff(payload: HookPayload, options: CliOptions) -> ReviewOutcome:
    """Open the diff review page and resolve to an outcome. Never raises."""
    cwd = payload.cwd or str(Path.cwd())
    log(f"session {payload.session_id} — reviewing {describe_scope(options.scope)} in {cwd}")

    repo = collect_diff(cwd, options.scope)
    # Seeded with the untracked files `collect_diff` refused: one count, one report.
    dropped_paths = repo.dropped_untracked

    def on_drop(_file: DiffFile) -> None:
        nonlocal dropped_paths
        dropped_paths += 1

    changed = parse_unified_diff(repo.unified, on_drop)
    files = filter_files(changed, options.scope)

    # An -I/-X pair matching nothing is otherwise an empty scope. Carried on the
    # outcome, because stderr does not reach an agent reading only `-o <file>`.
    filtered_out = len(changed) if changed and not files else 0
    if filtered_out:
        warn(
            f"every one of the {len(changed)} changed file(s) was removed by the path "
            f"filters — nothing is being reviewed in {repo.scope_label}"
        )
        warn("-I/--include and -X/--exclude prefixes are relative to the repository root")

    if not files:
        note = (
            f"No changes to review in {repo.scope_label}."
            if repo.is_repo
            else "Not a git repository — no diff available."
        )
        log(f"{note} Allowing.")
        return ReviewOutcome(
            summary=ReviewOutcomeSummary(
                review=None,
                files=files,
                scope=repo.scope_label,
                branch=repo.branch,
                note=note,
                is_repo=repo.is_repo,
                filtered_out=filtered_out,
                # Same reason as `filtered_out`: an empty diff is not always a
                # clean tree.
                untracked_scan_failed=repo.untracked_scan_failed,
                dropped_paths=dropped_paths,
            )
        )

    # Only where staging is meaningful: in a ref/range scope the index says
    # nothing about the reviewed content, and acting on it would reach outside it.
    can_stage = repo.is_repo and options.scope.kind in ("worktree", "staged")
    if can_stage:
        states = get_stage_states(cwd)
        for diff_file in files:
            diff_file.staged = states.get(diff_file.path, "no")

    # Unlike the empty branch above, a failed scan here leaves a review that
    # looks complete while every new file is missing. Say it on the page and in
    # the report.
    scan_warning = (
        "Listing untracked files failed — any new file in this scope is missing from this diff."
        if repo.untracked_scan_failed
        else None
    )
    if scan_warning:
        warn(scan_warning)

    ctx = ReviewContext(
        # The resolved cwd: the stage routes run git in it, and the raw one may
        # be empty.
        payload=HookPayload(
            session_id=payload.session_id,
            timestamp=payload.timestamp,
            cwd=cwd,
            tool_name=payload.tool_name,
            plan=payload.plan,
        ),
        branch=repo.branch,
        files=files,
        is_repo=repo.is_repo,
        can_stage=can_stage,
        mode="diff",
        scope=repo.scope_label,
        note=None if repo.is_repo else "Not a git repository — no diff available.",
        warning=scan_warning,
    )

    server = start_review_server(ctx)
    log(f"review page at {server.url}")
    log(
        f"{len(files)} file(s) changed — "
        f"{'opening browser…' if options.open else 'open it to review'}"
    )
    if options.open:
        open_browser(server.url)

    try:
        # ONLY this one statement is fail-open, and it catches only ServerClosed:
        # once a verdict has arrived, a later error must not be reported as "no
        # review captured" and invert it into an approval.
        try:
            review = server.gate.wait()
        except ServerClosed as err:
            note = f"No review was captured ({err})."
            warn(f"{note} Allowing.")
            return ReviewOutcome(
                summary=ReviewOutcomeSummary(
                    review=None,
                    files=files,
                    scope=repo.scope_label,
                    branch=repo.branch,
                    note=note,
                    interrupted=True,
                    is_repo=repo.is_repo,
                    untracked_scan_failed=repo.untracked_scan_failed,
                    dropped_paths=dropped_paths,
                )
            )

        # Persist first: the archive is what survives an agent that ignores the
        # report.
        save_history(
            review,
            files,
            HistoryMeta(
                cwd=cwd,
                session_id=payload.session_id,
                scope=repo.scope_label,
                branch=repo.branch,
                mode="diff",
                untracked_scan_failed=repo.untracked_scan_failed,
                dropped_paths=dropped_paths,
                enabled=options.history,
                history_dir=options.history_dir,
            ),
        )
        log(
            f"changes requested ({len(review.comments)} comment(s))"
            if review.decision == "request_changes"
            else "approved"
        )
        return ReviewOutcome(
            summary=ReviewOutcomeSummary(
                review=review,
                files=files,
                scope=repo.scope_label,
                branch=repo.branch,
                is_repo=repo.is_repo,
                # Carried even beside a verdict: it covers only the files that
                # got in.
                untracked_scan_failed=repo.untracked_scan_failed,
                dropped_paths=dropped_paths,
            )
        )
    finally:
        server.close()

File: src/cli/review-command.ts
// The `revgate review` command. NOT a hook: no payload on stdin, no hook JSON on
// stdout, real exit codes. The plan half reuses `gatePlan` from the hook module.
import { readFile, writeFile } from "node:fs/promises";
import path from "node:path";
import { collectDiff } from "../git/collect.js";
import { describeScope, filterFiles, ScopeError } from "../git/scope.js";
import { getStageStates } from "../git/staging.js";
import { parseUnifiedDiff } from "../review/diff.js";
import { reviewReport, type ReviewOutcomeSummary } from "../review/report.js";
import { openBrowser } from "../server/browser.js";
import { startReviewServer, type ReviewContext } from "../server/index.js";
import { log, warn } from "../shared/log.js";
import { saveHistory } from "../store/history.js";
import type { HookDecision, HookPayload, ReviewSubmission } from "../shared/types.js";
import type { CliOptions } from "./args.js";
import { gatePlan } from "./plan-hook.js";

/** One review's output: `ReviewOutcomeSummary` plus what only the hook paths need. */
export interface ReviewOutcome extends ReviewOutcomeSummary {
  /** The hook verdict, set only by the plan path. */
  decision?: HookDecision;
}

/**
 * The plan text to review, or null when `--plan` was not asked for: `--plan
 * <file>` beats `$REVGATE_PLAN_FILE`. Strict, because the skill reads exit 0 as
 * "approved, start implementing" — a missing plan is bad usage (exit 2).
 */
export async function resolvePlan(options: CliOptions, cwd: string): Promise<string | null> {
  if (!options.plan) return null;

  // `||`, not `??`: an empty planFile is "no path was given", and falling back
  // to the env var is exactly what bare `--plan` documents.
  const file = options.planFile || process.env.REVGATE_PLAN_FILE;
  if (file) {
    try {
      const text = await readFile(path.resolve(cwd, file), "utf8");
      // An existing-but-empty file is not a plan: reviewing a blank document the
      // reviewer can only approve is a sign-off on nothing.
      if (text.trim()) return text;
      warn(`plan file ${file} is empty`);
    } catch (err) {
      throw new ScopeError(`could not read plan file ${file}: ${(err as Error).message}`);
    }
  }
  throw new ScopeError(
    "--plan was given but no plan text was found — pass a file or set $REVGATE_PLAN_FILE",
  );
}

/** `revgate review` — the on-demand entry point the skill drives. */
export async function runReviewCommand(options: CliOptions): Promise<void> {
  const cwd = process.cwd();
  const payload: HookPayload = { sessionId: "cli", timestamp: Date.now(), cwd };

  const planText = await resolvePlan(options, cwd);
  const outcome = planText != null
    ? await gatePlan(payload, planText, options)
    : await reviewDiff(payload, options);

  // `--output` when asked, otherwise stdout: the only thing this command ever
  // writes there, and never hook JSON.
  const deliver = async (text: string) => {
    if (!options.output) {
      process.stdout.write(text);
      return;
    }
    const dest = path.resolve(cwd, options.output);
    try {
      await writeFile(dest, text, "utf8");
      log(`annotations written to ${dest}`);
    } catch (err) {
      // The human has already reviewed, and exit 1 reads as "no verdict".
      warn(`could not write ${dest}: ${(err as Error).message}`);
      warn("writing the annotations to stdout instead");
      process.stdout.write(text);
    }
  };

  const report = reviewReport(
    outcome,
    planText != null ? "plan" : "diff",
    options.exitCodeOnComments,
  );
  if (report.kind === "interrupted") {
    warn("no verdict was captured — reporting an error rather than an approval");
  } else if (report.kind === "not-a-repo") {
    warn("not a git repository — nothing was reviewed");
    warn("run `revgate review` from inside a repository");
  } else if (report.kind === "scan-failed") {
    warn("the untracked-file scan failed — reporting an error rather than an approval");
  } else if (report.kind === "dropped-paths") {
    warn("every changed file was dropped for a line break in its path — not an approval");
  }
  await deliver(report.text);
  process.exitCode = report.exitCode;
}

/** Open the diff review UI and resolve to an outcome. Never throws. */
export async function reviewDiff(
  payload: HookPayload,
  options: CliOptions,
): Promise<ReviewOutcome> {
  const cwd = payload.cwd || process.cwd();
  log(`session ${payload.sessionId} — reviewing ${describeScope(options.scope)} in ${cwd}`);

  const repo = await collectDiff(cwd, options.scope);
  // Seeded with the untracked files `collectDiff` refused: one count, one report.
  let droppedPaths = repo.droppedUntracked ?? 0;
  const changed = parseUnifiedDiff(repo.unified, () => {
    droppedPaths++;
  });
  const files = filterFiles(changed, options.scope);

  // An -I/-X pair matching nothing is otherwise an empty scope. Carried on the
  // outcome, because stderr does not reach an agent reading only `-o <file>`.
  const filteredOut = changed.length > 0 && files.length === 0 ? changed.length : 0;
  if (filteredOut) {
    warn(
      `every one of the ${changed.length} changed file(s) was removed by the path ` +
        `filters — nothing is being reviewed in ${repo.scopeLabel}`,
    );
    warn("-I/--include and -X/--exclude prefixes are relative to the repository root");
  }

  // No changes: nothing to review.
  if (files.length === 0) {
    const note = repo.isRepo
      ? `No changes to review in ${repo.scopeLabel}.`
      : "Not a git repository — no diff available.";
    log(`${note} Allowing.`);
    return {
      review: null,
      files,
      scope: repo.scopeLabel,
      branch: repo.branch,
      note,
      isRepo: repo.isRepo,
      filteredOut,
      // Same reason as `filteredOut`: an empty diff is not always a clean tree.
      untrackedScanFailed: repo.untrackedScanFailed,
      droppedPaths,
    };
  }

  // Only where staging is meaningful: in a ref/range scope the index says
  // nothing about the reviewed content, and acting on it would reach outside it.
  const scopeKind = options.scope.kind;
  const canStage = repo.isRepo && (scopeKind === "worktree" || scopeKind === "staged");
  if (canStage) {
    const states = await getStageStates(cwd);
    for (const f of files) f.staged = states[f.path] ?? "no";
  }

  // Unlike the empty branch above, a failed scan here leaves a review that looks
  // complete while every new file is missing. Say it on the page and the report.
  const scanWarning = repo.untrackedScanFailed
    ? "Listing untracked files failed — any new file in this scope is missing from this diff."
    : undefined;
  if (scanWarning) warn(scanWarning);

  const ctx: ReviewContext = {
    // The resolved cwd: the stage routes run git in it, and the raw one may be empty.
    payload: { ...payload, cwd },
    branch: repo.branch,
    files,
    isRepo: repo.isRepo,
    canStage,
    mode: "diff",
    scope: repo.scopeLabel,
    note: repo.isRepo ? undefined : "Not a git repository — no diff available.",
    warning: scanWarning,
  };

  const server = await startReviewServer(ctx);
  log(`review UI at ${server.url}`);
  log(`${files.length} file(s) changed — ${options.open ? "opening browser…" : "open it to review"}`);
  if (options.open) openBrowser(server.url);

  try {
    // ONLY the await is fail-open: once a verdict has arrived, a later throw must
    // not be reported as "no review captured" and invert it into an approval.
    let review: ReviewSubmission;
    try {
      review = await server.waitForSubmission;
    } catch (err) {
      const note = `No review was captured (${(err as Error).message}).`;
      warn(`${note} Allowing.`);
      return {
        review: null,
        files,
        scope: repo.scopeLabel,
        branch: repo.branch,
        note,
        interrupted: true,
        isRepo: repo.isRepo,
        untrackedScanFailed: repo.untrackedScanFailed,
        droppedPaths,
      };
    }

    // Persist first: the archive is what survives an agent that ignores the report.
    await saveHistory(review, files, {
      cwd,
      sessionId: payload.sessionId,
      scope: repo.scopeLabel,
      branch: repo.branch,
      mode: "diff",
      untrackedScanFailed: repo.untrackedScanFailed,
      droppedPaths,
      enabled: options.history,
      historyDir: options.historyDir,
    });
    log(
      review.decision === "request_changes"
        ? `changes requested (${review.comments.length} comment(s))`
        : "approved",
    );
    return {
      review,
      files,
      scope: repo.scopeLabel,
      branch: repo.branch,
      isRepo: repo.isRepo,
      // Carried even beside a verdict: it covers only the files that got in.
      untrackedScanFailed: repo.untrackedScanFailed,
      droppedPaths,
    };
  } finally {
    // Always: a throw after the submission would leave the listener open.
    server.close();
  }
}
